### PR TITLE
[SPARK-17475] [STREAMING] Delete CRC files if the filesystem doesn't use checksum files

### PR DESCRIPTION
**Pull request by frreiss**
## What changes were proposed in this pull request?

When the metadata logs for various parts of Structured Streaming are stored on non-HDFS filesystems such as NFS or ext4, the HDFSMetadataLog class leaves hidden HDFS-style checksum (CRC) files in the log directory, one file per batch. This PR modifies HDFSMetadataLog so that it detects the use of a filesystem that doesn't use CRC files and removes the CRC files.
## How was this patch tested?

Modified an existing test case in HDFSMetadataLogSuite to check whether HDFSMetadataLog correctly removes CRC files on the local POSIX filesystem.  Ran the entire regression suite.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala [hunk 1]
```diff
@@ -146,6 +146,11 @@ class HDFSMetadataLog[T: ClassTag](sparkSession: SparkSession, path: String)
           // It will fail if there is an existing file (someone has committed the batch)
           logDebug(s"Attempting to write log #${batchIdToPath(batchId)}")
           fileManager.rename(tempPath, batchIdToPath(batchId))
+
+          // SPARK-17475: HDFSMetadataLog should not leak CRC files
+          // If the underlying filesystem didn't rename the CRC file, delete it.
```
Comment (jodersky) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: Is this specific to streaming, or would other parts of spark benefit if this behavior were changed in the file manager?

Comment (frreiss) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: I believe HDFSMetadataLog is only called from Structured Streaming classes currently.

Comment (jodersky) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: ok, looks good then

**Comment by rxin** [comment]
Jenkins, test this please.

**Comment by zsxwing** [comment]
@frreiss  I saw LocalFileSystem will also rename the checksum file: https://github.com/apache/hadoop/blob/release-2.7.1/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/fs/LocalFileSystem.java#L137

So I think this fix is not needed.

**Comment by frreiss** [comment]
When I comment out line 155 in HDFSMetadataLog.scala on this branch (`if (fileManager.exists(crcPath)) fileManager.delete(crcPath)`) and run the test case attached to this PR, the test case fails:
```
[freiss@fuzzy]:~/spark/from_git/spark-17475$ build/sbt -Phadoop-2.7 -Pscala-2.11 "test-only org.apache.spark.sql.execution.streaming.HDFSMetadataLogSuite"
...
[info] - HDFSMetadataLog: purge *** FAILED *** (135 milliseconds)
[info]   Array(/Users/freiss/spark/from_git/spark-17475/target/tmp/spark-682aa8da-3f04-494f-846f-13c97d3e5538/..29ef67f7-1712-4350-8552-1f8bc6424d0b.tmp.crc, /Users/freiss/spark/from_git/spark-17475/target/tmp/spark-682aa8da-3f04-494f-846f-13c97d3e5538/..ab9bafcb-bdf5-4411-9a9b-60d293d653a6.tmp.crc, /Users/freiss/spark/from_git/spark-17475/target/tmp/spark-682aa8da-3f04-494f-846f-13c97d3e5538/..f79fbc34-58c5-4856-a40d-84eef49c8b9e.tmp.crc, /Users/freiss/spark/from_git/spark-17475/target/tmp/spark-682aa8da-3f04-494f-846f-13c97d3e5538/2) had size 4 instead of expected size 1 (HDFSMetadataLogSuite.scala:126)
[info]   org.scalatest.exceptions.TestFailedException:
[info]   at org.scalatest.Assertions$class.newAssertionFailedException(Assertions.scala:500)
...
```
Note how the test case failure message contains a list of orphan .crc files left in HDFSMetadataLog's temp directory. So, while the filesystem code appears on the surface to be immune to this problem, the problem is clearly happening in the context of the unit tests. Determining exactly what is going on will require more in-depth investigation. Depending on the true root cause, it's possible that this problem also occurs in some distributed settings.

**Comment by zsxwing** [comment]
retest this please

**Comment by zsxwing** [comment]
It turns out there is a bug in Hadoop's FileContext that doesn't rename the checksum file.

LGTM pending tests.

**Comment by viirya** [comment]
Should we just delete the crc file or rename it? As the crc file is created for the temp file, I suppose the filesystem uses it?

**Comment by rxin** [comment]
Merging in master/branch-2.1.

**Comment by frreiss** [comment]
@viirya to answer your question re deleting vs moving the files: Deleting is easier to implement, because once the .crc file is deleted, you can be sure it won't appear again. Moving the checksum files would be better but would require tracking down other Spark code that does filesystem operations on the files later on. These log files are very short by HDFS standards, so the checksums aren't providing much value anyway.

**Comment by viirya** [comment]
@frreiss Thanks for explanation!

**Pull request closed without merging**